src/jdk.incubator.code/share/classes/jdk/incubator/code/op/CoreOp.java

```diff
@@ -4435,4 +4435,85 @@ public static BinaryTestOp le(Value lhs, Value rhs) {
     public static ConcatOp concat(Value lhs, Value rhs) {
         return new ConcatOp(lhs, rhs);
     }
+
+    public static FuncOp quoteOp(Op op) {
+
+        return func("q", FunctionType.VOID).body(block -> {
+            List<Value> usedValues = new ArrayList<>();
+            usedValues.addAll(op.capturedValues());
+            usedValues.addAll(op.operands());
+            for (Value v : usedValues) {
+                TypeElement t;
+                if (v instanceof Op.Result opr && opr.op() instanceof VarOp varOp) {
+                    t = varOp.varValueType();
+                } else {
+                    t = v.type();
+                }
+                Block.Parameter p = block.parameter(t);
+                Op.Result nv = block.op(var(p));
+                block.context().mapValue(v, nv);
+            }
+
+            Body.Builder qbody = Body.Builder.of(block.parentBody(), FunctionType.VOID, CopyContext.create(block.context()));
+            Block.Builder qblock = qbody.entryBlock();
+
+            if (op.ancestorBody() != null) {
+                // needed for reachability check (of captured values) to succeed
+                qblock.context().mapBlock(op.ancestorBody().entryBlock(), qblock);
+            }
+            Op.Result opr = qblock.op(op, (b, o) -> {
+                // a modified copy transformer, that insert var.load before use of captured values (if needed)
+                Map<Value, Value> m = new HashMap<>();
+                for (Value operand : o.operands()) {
+                    if (usedValues.contains(operand) && !(operand instanceof Op.Result r && r.op() instanceof VarOp)) {
+                        Value capVar = b.context().getValue(operand);
+                        Op.Result capVal = b.op(varLoad(capVar));
+                        m.put(operand, capVar);
+                        b.context().mapValue(operand, capVal);
+                    }
+                }
+
+                b.op(o);
+
+                for (Value k : m.keySet()) {
```

Comment:
Why is this loop needed? Is this to restore the original mapping in the copy context? If so, perhaps the code could be strengthened by surrounding `b.op` with a `try/finally` and then restore the mapping in the finally block?

(But I wonder if the API could offer more help here -- e.g. some kind of temp value mapping)